### PR TITLE
fix label smoothing in focal loss

### LINKED ISSUE #672: Possibly wrong implementation of focal loss in regard of label smoothing
I think that there is a wrong implementation of focal loss in regard of label smoothing.

Here is the current implementation
https://github.com/keras-team/keras-cv/blob/35b9f9534fc193e00b4fa850963a05b331467065/keras_cv/losses/focal.py#L74-L90

as seen above, label smoothing occurs *before* the computation of `alpha` and `pt`. From the original paper, the value expected at this stage for `y_true` should be either `0` or `1`. However after label smoothing, `y_true` will be in `]0,1[`, therefore invalidating `alpha` and `pt` computation (at least not as written in the original paper).

Also to compute Multi-Label, I think the final reduction should be a sum to be correct. One alternative would be, as implemented in pytorch to add a `reduction` parameter, to either:

* 'none': perform no class reduction
* 'sum': perform a sum over classes
* 'mean': perform a mean over classes

What do you thin @LukeWood , @quantumalaviya ?

### PR DESCRIPTION
**Pull request by DavidLandup0**
# What does this PR do?

Introduced in #672, PR'd in #680 and re-opened due to seemingly going cold.
If @atuleu still wants to do this PR, feel free to ping here. Otherwise, this is a bug that should be fixed as the object detection API is being built.

Added numerical tests to verify that the loss is indeed `0.0` on a fully correct prediction, with label smoothing on, which was off in the original implementation.

Fixes #672

## Who can review?
@atuleu @LukeWood @tanzhenyu

### CODE REVIEW

**Review by LukeWood — APPROVED** [review]

**Comment by ianstenbit** [comment]
@DavidLandup0 lmk if you'd like to rebase and merge this, otherwise I'll close it.

Thanks for the PRs, and sorry that many of these got buried.

**Comment by ianstenbit** [comment]
Oops -- didn't mean to close it just yet

**Comment by DavidLandup0** [comment]
Yep, I'll rebase this :)
Thanks for taking another look! It changes the numerics of the loss function at `beta` != 1. Should we use `1` as the default to ensure backward compatability, unless we already do?

**Comment by ianstenbit** [comment]
> Yep, I'll rebase this :) Thanks for taking another look! It changes the numerics of the loss function at `beta` != 1. Should we use `1` as the default to ensure backward compatability, unless we already do?

I'm not super familiar with this loss, so I'm not sure.

It seems like the only delta here is that we do the label smoothing after computing `pt` instead of before -- is there some other nuance I'm missing?

I don't think there is wide usage of label smoothing with this, so if the old version is not quite numerically correct I'm not concerned with making that reproducible, so backwards compat isn't a huge issue unless it's an API breakage.
